debug/custom-vector-source.html

```diff
@@ -0,0 +1,79 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>Mapbox GL JS debug page</title>
+    <meta charset='utf-8'>
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
+    <link rel='stylesheet' href='../dist/mapbox-gl.css' />
+    <style>
+        body { margin: 0; padding: 0; }
+        html, body, #map { height: 100%; }
+    </style>
+</head>
+
+<body>
+<div id='map'></div>
+
+<script src='../dist/mapbox-gl-dev.js'></script>
+<script src='../debug/access_token_generated.js'></script>
+<script type="module">
+import Protobuf from 'https://cdn.skypack.dev/pbf';
+import vt from 'https://cdn.skypack.dev/@mapbox/vector-tile';
+
+var map = window.map = new mapboxgl.Map({
+    container: 'map',
+    zoom: 12,
+    center: [-87.622088, 41.878781],
+    style: 'mapbox://styles/mapbox/light-v10',
+    hash: true
+});
+
+class CustomVectorSource {
+    constructor() {
+        this.id = 'custom-vector';
+        this.type = 'custom';
+        this.minzoom = 6;
+        this.maxzoom = 14;
+        this.tileSize = 512;
+        this.tilesUrl = 'https://tiles.mapillary.com/maps/vtp/mly1_public/2/{z}/{x}/{y}?access_token=MLY|4142433049200173|72206abe5035850d6743b23a49c41333';
+    }
+
+    async loadTile(tile, {signal}) {
+        const url = this.tilesUrl
+            .replace('{z}', String(tile.z))
+            .replace('{x}', String(tile.x))
+            .replace('{y}', String(tile.y));
+
+        const response = await fetch(url, {signal});
+        const data = await response.arrayBuffer();
+        const vectorTile = new vt.VectorTile(new Protobuf(data));
+
+        return vectorTile;
+    }
+}
```

Comment:
I apologize if my question sent you down a wrong path, but I'm wondering if this example shows that CustomVectorSource isn't necessary. I talked to @ansis during standup this morning and he noted that there aren't many great tools for creating vector tiles and it's not a format many know how to work with. Plus, we can't really animate vector tiles performantly. If we consider this interface as being a way to allow rendering and animation of tile-able sources that aren't currently possible in GL JS, I think maybe we do come back to the idea that raster sources are the only thing that really make sense.

Also, looking at these examples, I think maybe I like the interface object style better than writing a class. We expect sources to be objects already.

Reply:
No worries, I just wondered about the other applications of the `CustomSource`. I'll drop this example since we didn't intend to implement this at the moment.

We can still have both notions to create a custom source. However, it would be more convenient to access the source methods like `coveringTiles` and `update` on the instance than pull the source with `map.getSource`.

```js
map.addSource('custom-raster', {
    type: 'custom',
    tileSize: 256,
    tilesUrl: 'https://stamen-tiles.a.ssl.fastly.net/watercolor/{z}/{x}/{y}.jpg',
    async loadTile(tile, {signal}) {}
});

const source = map.getSource('custom-raster');
const coveringTiles = source.coveringTiles();
```

```js
class CustomRasterSource {
    constructor() {
        this.id = 'custom-raster';
        this.type = 'custom';
        this.tileSize = 256;
        this.tilesUrl = 'https://stamen-tiles.a.ssl.fastly.net/watercolor/{z}/{x}/{y}.jpg';
    }

    async loadTile(tile, {signal}) {}
}

const source = new CustomRasterSource();

map.addSource('custom-raster', source);
const coveringTiles = source.coveringTiles();
```